### PR TITLE
Support PDF Display in UIWebViews

### PR DESCRIPTION
**Pull request by keefertaylor**
#704 
In a webview with HTML content, there is a valid view at _webView._internal.browserView, but the view is always empty. This causes EG's UIWebView Idling Resource to assume the webview is waiting to render.

In a webview with a PDF, there is a view at _webView._internal.pdfHandler.pdfView. This view *only* exists after the PDF is rendered (AFAICT) and does not exist if the webview is rendering something other than a PDF.

To fix, check if there is a rendered PDF in the webview before performing HTML rendering checks.

### CODE REVIEW

File: EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m [hunk 1]
```diff
@@ -137,6 +137,11 @@ - (BOOL)isIdleNow {
     }
   }
 
+  // UIWebViews may be used to display PDFs in addition to HTML based content. Test if there's a PDF
+  // view populated in UIWebView's hierarchy.
+  BOOL webViewIsDisplayingPDF = ([[[_webView valueForKey:@"_internal"] valueForKey:@"pdfHandler"]
```
[the remark this answers is not in the document]
Comment (keefertaylor) [reply EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m]: > Do we have to do something with the bool?
Yup, need to return that we are idle.

> Can we add a test case with a PDF to prevent regressions & prove the issue is fixed?
Sure, let me put something together.

[the remark this answers is not in the document]
Comment (keefertaylor) [reply EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m]: I wired up a basic test case that shows that opening a PDF will cause the app to never reach an idle state. PTAL.

[the remark this answers is not in the document]
Comment (bootstraponline) [reply EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m]: Nice, looks good to me. Someone on the EarlGrey will need to approve before the PR is merged.

**Comment by keefertaylor** [comment]
Adding a comment to get CLABot to re-run.

File: EarlGrey.xcodeproj/xcshareddata/xcschemes/EarlGrey.xcscheme [hunk 2]
```diff
@@ -26,16 +26,37 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      language = ""
```
Comment (khandpur) [inline EarlGrey.xcodeproj/xcshareddata/xcschemes/EarlGrey.xcscheme]: Are these changes intentional or just a side-effect of opening in a different Xcode?

File: EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m [hunk 3]
```diff
@@ -137,6 +137,14 @@ - (BOOL)isIdleNow {
...
+                                     valueForKey:@"pdfView"] == nil);
+  if (webViewIsDisplayingPDF) {
```
Comment (khandpur) [inline EarlGrey/Synchronization/GREYUIWebViewIdlingResource.m]: It's not clear to me what's throwing off EarlGrey sync logic.

@tirodkar Can you please see if it's something we can fix in the core-logic rather than short circuiting the entire method when PDFs are being loaded?

File: Tests/FunctionalTests/Sources/FTRUIWebViewTest.m [hunk 4]
```diff
@@ -108,6 +108,13 @@ - (void)DISABLED_testSynchronizationWhenSwitchingBetweenLoadingMethods {
       assertWithMatcher:grey_sufficientlyVisible()];
 }
 
+- (void)testPDFIdlesCorrectly {
```
Comment (khandpur) [inline Tests/FunctionalTests/Sources/FTRUIWebViewTest.m]: I'd recommend splitting this up with local only and network-only scenarios (for an internal use-case)

File: Tests/FunctionalTests/TestRig/Sources/FTRWebViewController.m [hunk 5]
```diff
@@ -77,6 +84,12 @@ - (IBAction)userDidTapLoadHTMLUsingLoadHTMLString {
   _webView.scrollView.bounces = _webViewBounceSwitch.isOn;
 }
 
+- (IBAction)userDidTapLoadPDF {
+  NSData *pdfData = [NSData dataWithContentsOfURL:[[self class] urlToTestPDF]];
+  [_webView loadData:pdfData MIMEType:@"application/pdf" textEncodingName:@"utf-8"
+   baseURL:[NSURL URLWithString: @"about:blank"]];
```
Comment (khandpur) [inline Tests/FunctionalTests/TestRig/Sources/FTRWebViewController.m]: formatting seems off.

**Comment by tirodkar** [comment]
Hey @keefertaylor, we're getting this change in as part of our internal to external code sync with any changes that were requested here. I'll close this for now.
